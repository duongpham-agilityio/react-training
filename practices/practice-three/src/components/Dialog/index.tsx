import {
  AlertDialog,
  AlertDialogBody,
  AlertDialogCloseButton,
  AlertDialogContent,
  AlertDialogFooter,
  AlertDialogHeader,
  AlertDialogOverlay,
  Button,
} from '@chakra-ui/react';
import { memo, useRef } from 'react';

export type TDialogProps = {
  title: string;
  description?: string;
  isOpen: boolean;
  onClose: () => void;
  onAccept: () => void;
  onCancel?: () => void;
};

const DialogComponent = (props: TDialogProps) => {
  const { title, description, isOpen, onClose, onAccept, onCancel } = props;
  const cancelRef = useRef<HTMLButtonElement>(null);

  return (
    <AlertDialog
      motionPreset="slideInBottom"
      isCentered
      leastDestructiveRef={cancelRef}
      isOpen={isOpen}
      onClose={onClose}
    >
      <AlertDialogOverlay />

      <AlertDialogContent>
        <AlertDialogHeader>{title}</AlertDialogHeader>
        <AlertDialogCloseButton />

        {description && <AlertDialogBody>{description}</AlertDialogBody>}

        <AlertDialogFooter>
          <Button
            ref={cancelRef}
            onClick={onCancel ?? onClose}
            color="darkGray30"
            bg="darkToLight"
          >
            No
          </Button>
          <Button colorScheme="red" ml={3} onClick={onAccept}>
            Yes
          </Button>
        </AlertDialogFooter>
      </AlertDialogContent>
    </AlertDialog>
  );
};

export default memo(DialogComponent);
