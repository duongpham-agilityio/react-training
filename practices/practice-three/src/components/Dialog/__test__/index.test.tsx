import { fireEvent, render } from '@testing-library/react';

// Components
import Dialog, { TDialogProps } from '..';

const props: TDialogProps = {
  isOpen: true,
  title: 'Test',
  description: 'Implement unit test for Dialog',
  onAccept: jest.fn(),
  onClose: jest.fn(),
};

const setup = () => render(<Dialog {...props} />);

describe('Dialog', () => {
  it('Match to snapshot', () => {
    const { container } = setup();

    expect(container).toMatchSnapshot();
  });

  it('onClose', () => {
    const { getByText } = setup();

    fireEvent.click(getByText('No'));
    expect(props.onClose).toBeCalled();
  });

  it('onAccept', () => {
    const { getByText } = setup();

    fireEvent.click(getByText('Yes'));
    expect(props.onAccept).toBeCalled();
  });
});
