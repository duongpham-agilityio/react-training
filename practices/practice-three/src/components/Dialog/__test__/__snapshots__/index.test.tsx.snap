// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog Match to snapshot 1`] = `
<div
  aria-hidden="true"
  data-aria-hidden="true"
/>
`;
